### PR TITLE
Fix ternaryOp program hash collision to account for logical dim when per-operand broadcasts are different but alias to same key

### LINKED ISSUE #43368: ttnn.where op (and eltwise ternary in general) has broken program cache key which will incorrectly alias kernels where input operands have different broadcasting requirements
fix tbd -> add per-operand broadcast information into the cache key instead of aliasing it all under ROW_COL_BROADCAST + volumes

### PR DESCRIPTION
**Pull request by jameszianxuTT**
### Summary
Closes #43368
Related https://github.com/tenstorrent/tt-xla/issues/4418

When executing multiple ttnn.where ops in sequence with different per-operand broacasting, the wrong cached program gets used due to incorrect cache key generation.

For tiled tensors, which are padded to 32x32 in the back 2 dims, two tensors with different logical shapes could alias to the same padded shape. However, the logical shape determines which broadcasting kernel to use, so there is incompete information. 

eg. A=1x8x1, B=1x1x32 are both padded to 1x32x32, and could classify as the same overall ROW_COL_BCAST type, but if they are broadcast to C=1x4x32, A needs both row and column broadcast, while B needs only a row broadcast. 

### Notes for reviewers
This bug originated from forge DS4 bringup, where the lowering of a pattern of torch tensor slice + mutations turned into 4 ttnn.where ops, which required different per-operand broadcasting and corrupted the output. This failure went away when program cache was disabled and was eventually root caused to this program hash collision bug.

Added test `pytest -svv tests/ttnn/unit_tests/operations/eltwise/test_where.py::test_where_program_cache_different_broadcast_shapes`

More details in  https://github.com/tenstorrent/tt-xla/issues/4418

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:jzx/repro_slice_bug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:jzx/repro_slice_bug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:jzx/repro_slice_bug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:jzx/repro_slice_bug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:jzx/repro_slice_bug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:jzx/repro_slice_bug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:jzx/repro_slice_bug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:jzx/repro_slice_bug)

### CODE REVIEW

File: tests/ttnn/unit_tests/operations/eltwise/test_where.py [hunk 1]
```diff
@@ -1075,3 +1075,79 @@ def test_where_ttt_int32_predicate_sub_core_grid(device, shape, sub_core_grid, c
...
+    # First where: predicate broadcasts on rows (1x32 -> 4x32)
+    pred1 = torch.zeros(1, 1, 32, dtype=torch.bfloat16)
+    pred1[:, :, 16:] = 1.0
+
+    # Second where: predicate broadcasts on columns (8x1 -> 8x32)
+    pred2 = torch.zeros(1, 8, 1, dtype=torch.bfloat16)
+    pred2[:, 4:, :] = 1.0
+
+    if variant == "TTT":
+        true1 = torch.randn(1, 1, 1, dtype=torch.bfloat16)
```
Comment (jameszianxuTT) [inline tests/ttnn/unit_tests/operations/eltwise/test_where.py:1102]: The TTT parameterization would fail on main

**Comment by jameszianxuTT** [comment]
tbd: does this class of bug exist for eltwise binary ops as well?

**Review by dchenTT — APPROVED** [review]
It seems out of date that ternary still hash on tensor volume.

**Comment by jameszianxuTT** [comment]
New tests are run here: https://github.com/tenstorrent/tt-metal/actions/runs/25144069737/job/73831916370#step:7:8536

This is a defensive change, as it only expands the cache key to cover a logical gap.
